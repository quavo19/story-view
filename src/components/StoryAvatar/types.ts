import type {
  ImageProps,
  ImageStyle,
  PressableProps,
  TextProps,
  TextStyle,
  ViewStyle,
} from 'react-native';
import type { StoriesType } from '../StoryView/types';

export interface StoryAvatarProps extends StoryAvatarStyleProps {
  index: number;
  item?: StoriesType;
  pressedIndex?: number;
  isStoryViewVisible?: boolean;
  viewedStories?: Array<boolean[]>;
  openStories?: (position: number) => void;
}

export interface StoryAvatarStyleProps {
  userNameStyle?: TextStyle;
  userImageStyle?: ImageStyle;
  containerStyle?: ViewStyle;
  userImageProps?: ImageProps;
  userNameProps?: TextProps;
  rootProps?: PressableProps;
  viewedStoryContainerStyle?: ViewStyle;
}

export interface CircleAnimationProps {
  index: number;
  pressedIndex?: number;
  isStoryViewVisible?: boolean;
}
