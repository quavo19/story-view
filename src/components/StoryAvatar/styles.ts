import { StyleSheet } from 'react-native';
import { Colors, moderateScale } from '../../theme';

const styles = StyleSheet.create({
  image: {
    height: '95%',
    width: '95%',
    borderRadius: moderateScale(35),
  },
  imageContainer: {
    height: moderateScale(70),
    width: moderateScale(70),
    overflow: 'hidden',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: moderateScale(35),
    borderColor: Colors.pink,
    borderWidth: 2,
    marginRight: moderateScale(16),
  },
  viewedStoryContainer: {
    borderColor: Colors.inActiveColor,
  },
  username: {
    alignSelf: 'center',
    marginRight: moderateScale(16),
    marginTop: moderateScale(6),
    color: Colors.white,
    fontWeight: '500',
  },
});

export default styles;
