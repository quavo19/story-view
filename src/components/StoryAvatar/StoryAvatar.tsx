import React from 'react';
import { Pressable, StyleSheet, Text, View } from 'react-native';
import Animated from 'react-native-reanimated';
import useCircleAnimation from './hooks/useCircleAnimation';
import styles from './styles';
import type { StoryAvatarProps } from './types';

const StoryAvatar = ({
  item,
  index,
  pressedIndex,
  isStoryViewVisible,
  openStories,
  viewedStories = [],
  userNameStyle,
  userImageStyle,
  userImageProps,
  viewedStoryContainerStyle,
  userNameProps,
  rootProps,
  containerStyle,
}: StoryAvatarProps) => {
  const isUserStorySeen: boolean = viewedStories?.[index]?.every(
    (val: boolean) => val
  );

  const _userNameStyle = StyleSheet.flatten([styles.username, userNameStyle]);
  const _userImageStyle = StyleSheet.flatten([styles.image, userImageStyle]);
  const _containerStyle = StyleSheet.flatten([
    styles.imageContainer,
    containerStyle,
    (isUserStorySeen && viewedStoryContainerStyle) ??
      styles.viewedStoryContainer,
  ]);

  const { avatarAnimatedStyle } = useCircleAnimation({
    pressedIndex,
    index,
    isStoryViewVisible,
  });

  return (
    <Pressable onPress={() => openStories?.(index)} {...rootProps}>
      <View style={_containerStyle}>
        <Animated.Image
          resizeMode="cover"
          source={{ uri: item?.profile }}
          style={[_userImageStyle, avatarAnimatedStyle]}
          {...userImageProps}
        />
      </View>
      <Text style={_userNameStyle} {...userNameProps}>
        {item?.username}
      </Text>
    </Pressable>
  );
};
export default StoryAvatar;
