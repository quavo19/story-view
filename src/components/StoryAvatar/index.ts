import StoryAvatar from './StoryAvatar';

export type { StoryAvatarProps } from './types';
export { StoryAvatar };
