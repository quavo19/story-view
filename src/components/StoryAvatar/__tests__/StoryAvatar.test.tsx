import React from 'react';
import { render } from '@testing-library/react-native';
import StoryAvatar from '../StoryAvatar';

jest.useFakeTimers();

describe('StoryAvatar Component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(<StoryAvatar index={0} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
