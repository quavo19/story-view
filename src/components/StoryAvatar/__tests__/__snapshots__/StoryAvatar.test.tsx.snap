// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryAvatar Component Match Snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#FFFFFF7F",
        "borderRadius": 52.5,
        "borderWidth": 2,
        "height": 105,
        "justifyContent": "center",
        "marginRight": 24,
        "overflow": "hidden",
        "width": 105,
      }
    }
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Object {
          "borderRadius": 52.5,
          "height": "95%",
          "width": "95%",
        }
      }
    />
  </View>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "#ffffff",
        "fontWeight": "500",
        "marginRight": 24,
        "marginTop": 9,
      }
    }
  />
</View>
`;
